### PR TITLE
`Development`: Avoid unnecessary REST calls for practice participation by loading them in the for-dashboard call

### PR DESCRIPTION
**Pull request by JohannesStoehr**
#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.cit.tum.de/dev/guidelines/development-process/#naming-conventions-for-github-pull-requests).
#### Server
- [x] **Important**: I implemented the changes with a very good performance and prevented too many (unnecessary) database calls.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/server/).
- [x] I added multiple integration tests (Spring) related to the features (with a high test coverage).
- [x] I documented the Java code using JavaDoc style.
#### Client
- [x] **Important**: I implemented the changes with a very good performance, prevented too many (unnecessary) REST calls and made sure the UI is responsive, even with large data.
- [x] I followed the [coding and design guidelines](https://docs.artemis.cit.tum.de/dev/guidelines/client/).

### Motivation and Context

Improve the performance of the exercise overview

### Description

Each time a row was extended in the exercise overview, the exercise details were loaded again to find the participations of the exercises. The loading of the participations is now taken care of in the for-dashboard call and removed, so it only has to be loaded once.
Since the practice participations are only needed in the specific courses, they are not included in the general for-dashboard call.

### Steps for Testing

Prerequisites:
- 1 Student
- 1 Programming Exercise
- 1 Modeling Exercise
- 1 Quiz Exercise

1. Log in to Artemis
2. Navigate to the exercise overview and check that all exercises, their scores and their action buttons are displayed correctly.
3. Check that there is no "details" call when expanding a row in the exercise overview
4. Wait after the due date of the programming exercise (or set it now to the past)
5. Use the practice mode in the programming exercise and submit a new partial correct solution
6. Check that the correct student actions (only cloning/online editor) are shown in the exercise overview even when reloading the page
7. Check that the score in the exercise row still shows the graded score

#### Performance Review
- [ ] I confirm that the client changes (in particular related to REST calls and UI responsiveness) are implemented with a very good performance 
- [ ] I confirm that the server changes (in particular related to database calls) are implemented with a very good performance
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

| Class/File | Line Coverage | Confirmation (assert/expect) |
|------------|--------------:|-----------------------------:|
| ProgrammingExercise.java | 92% | ✅                           |

### CODE REVIEW

File: src/main/webapp/app/overview/course-exercises/course-exercise-row.component.ts [hunk 1]
```diff
@@ -50,15 +50,8 @@ export class CourseExerciseRowComponent implements OnInit, OnDestroy, OnChanges
...
-            this.exerciseService.getExerciseDetails(this.exercise.id).subscribe((exerciseResponse: HttpResponse<Exercise>) => {
-                if (exerciseResponse.body) {
-                    this.exercise = exerciseResponse.body!;
-                    if (this.exercise.studentParticipations?.length) {
-                        this.gradedStudentParticipation = this.participationService.getSpecificStudentParticipation(this.exercise.studentParticipations, false);
-                    }
-                }
-            });
+        if (this.exercise?.studentParticipations?.length) {
+            this.gradedStudentParticipation = this.participationService.getSpecificStudentParticipation(this.exercise.studentParticipations, false);
```
Comment (bassner) [inline src/main/webapp/app/overview/course-exercises/course-exercise-row.component.ts:54]: Caused unused import of HttpResponse that was previously used here

**Review by TimOrtel — DISMISSED** [review]
The code looks good to me.

**Review by MaximilianAnzinger — CHANGES_REQUESTED** [review]

**Review by MaximilianAnzinger — APPROVED** [review]
Code LGTM
Manual LGTM

**Review by YeetTheFirst21 — APPROVED** [review]
Tested on TS2, there seems to be no details call when results are expanded or on any quiz and modelling exercise.

Only on programming exercises, a details call is given when the buildlog is requested which seems like intended to me:

Screenshot 2023-05-04 124057

**Review by pellumbbaboci — DISMISSED** [review]
Manual: Tested on ts1 and its working as expected

**Review by YeetTheFirst21 — DISMISSED** [review]
Tested again on TS1, all seems to be working including the new test cases

Screenshot 2023-05-05 123451

**Review by MaximilianAnzinger — DISMISSED** [review]
Tested the newly added steps (4-7) on local Atlassian setup. LGTM 👍

**Review by Hialus — APPROVED** [review]
Tested in testing session on TS1.
Code also looks good to me 👍🏻

**Review by DominikRemo — APPROVED** [review]
Tested in testing session on ts1. Works for me 👍

**Review by Hialus — APPROVED** [review]

**Review by pellumbbaboci — APPROVED** [review]
Tested during testing session on ts1 and works as expected

**Review by tobias-lippert — APPROVED** [review]
Tested in testing session on ts1. Nothing seems to be broken and the details endpoint is not called when the exercise row is extended

**Review by pal03377 — APPROVED** [review]
tested on ts1, works as expected

File: src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java [hunk 2]
```diff
@@ -572,6 +572,17 @@
...
+    @Query("""
+            SELECT DISTINCT p FROM StudentParticipation p
+            LEFT JOIN FETCH p.submissions s
+            LEFT JOIN FETCH s.results r
+            WHERE p.student.id = :studentId
+                AND p.exercise in :exercises
+                AND (p.testRun = FALSE OR :includeTestRuns = TRUE)
+            """)
+    List<StudentParticipation> findByStudentIdAndIndividualExercisesWithEagerSubmissionsResult(@Param("studentId") Long studentId, @Param("exercises") List<Exercise> exercises,
+            @Param("includeTestRuns") boolean includeTestRuns);
```
[the remark this answers is not in the document]
Comment (krusche) [reply src/main/java/de/tum/in/www1/artemis/repository/StudentParticipationRepository.java:584]: I suggest we work on this in a follow up

**Pull request merged**